[Accelerate] [vImage] Swift Overlays

FlexMonkey opened the pull request:
A suite of functions, enumerations, and option sets to make working with vImage in Swift simpler.

* vImage_Buffer - new initializers to instantiate and initialize buffers with a single function call.
* vImage_Buffer - new functions to copy buffers and create CGImage instances from contents.
* vImage_CGImageFormat - new initializers.
* vImageConverter - methods to wrap free functions.
* vImageConverter - new make and convert functions.
* vImageCVImageFormat - new make functions.
* vImageCVImageFormat - methods to wrap free functions.
* vImage_Error - errorDescription function.
* vImage flags as an option set.

cc: @moiseev @airspeedswift @stephentyrone

FlexMonkey committed:
Accelerate vImage Swift Overlays

A suite of functions, enumerations, and option sets to make working with vImage in Swift simpler.

* vImage_Buffer - new initializers to instantiate and initialize buffers with a single function call.
* vImage_Buffer - new functions to copy buffers and create CGImage instances from contents.
* vImage_CGImageFormat - new initializers.
* vImage_CGImageFormat - new equivalence operator.
* vImageConverter - methods to wrap free functions.
* vImageConverter - new make and convert functions.
* vImageCVImageFormat - new make functions.
* vImageCVImageFormat - methods to wrap free functions.
* vImage_Error - errorDescription function.
* vImage flags as an option set.

FlexMonkey committed:
Add new methods for generating CV -> CG and CG -> CV converters.

FlexMonkey committed:
Merge branch 'master' into accelerate-vImage

FlexMonkey committed:
update comments: `height` and `width` to `size`.

stephentyrone commented on stdlib/public/Darwin/Accelerate/vImage_Buffer.swift:
```diff
@@ -0,0 +1,348 @@
+extension vImage_Buffer {
+    
+    /// Returns the size, in pixels, of a vImage buffer.
+    public var size: CGSize {
```
Why do we want to expose a `size` property instead of the existing `.width` and `.height`? What's the intended use case for this?

FlexMonkey replied:
As per the header. I'm exposing the various free functions as computed properties.

 *  Rationale: If you attempt your own home-made conversion to CGSize / NSSize by
 *            ordinary C rules and the value rounds, it will round up half the time.
 *            This could lead to a crash later because the height or width
 *            will be reported to be larger than it really is and an ensuing image
 *            operation will attempt to touch scanlines that don't exist.

stephentyrone replied:
Ok, then please update the comment to be clear that the size is *rounded down* and you may lose part of your image, especially on 32b systems.

FlexMonkey replied:
done!

stephentyrone commented on stdlib/public/Darwin/Accelerate/vImage_Buffer.swift:
```diff
@@ -0,0 +1,348 @@
+    /// - Parameter alignment: The preferred alignment is written to this parameter.
+    ///
+    /// - Returns: An initialized vImage buffer.
+    public init?(size: CGSize,
```
It's *really* not OK to have an `init` function that leaves `self` in an uninitialized state. This needs re-working. What pattern are you trying to enable by providing this?

FlexMonkey replied:
But it does have `self.init()` (I've just moved it to after the size check). This initializer initializes a buffer without memory allocation (i.e. with `kvImageNoAllocate`) and writes the alignment to the `alignment` inout parameter.

stephentyrone replied:
The problem is that after calling this, the resulting vImage_Buffer object still isn't initialized. If the goal is to get the desired rowBytes and alignment, let's have a static function that *just* does that, rather than lying about it being an initializer.

FlexMonkey replied:
Good point. I've removed that faux initializer and replaced with a new static function: `preferredAlignmentAndRowBytes`.

stephentyrone commented on stdlib/public/Darwin/Accelerate/vImage_Buffer.swift:
```diff
@@ -0,0 +1,348 @@
+    /// - Parameter options: The options to use when performing this operation.
+    ///
+    /// - Returns: An initialized vImage buffer.
+    public init?(size: CGSize,
```
Why are we taking `CGSize` params for these functions? `CGSize` uses `CGFloat` to represent `width` and `height`, which causes problems due to rounding on 32b platforms (watchOS). Unless you have specific use cases in mind, it seems vastly preferable to use `Int` `width` and `height` parameters.

FlexMonkey replied:
My thinking was that the native `width` and `height` properties of a `vImage_Buffer` are `vImagePixelCount` (`UInt`) - it seemed a bit disjointed to have an integer parameter on the initializer, but the property exposed as a different type. Another point was that all existing vImage functions have `height` preceding `width` and that using `CGSize` might assist in avoiding errors with my `width` preceding `height`. 

That said, other image classes do use integer width and height, so I'll update accordingly.

stephentyrone commented on stdlib/public/Darwin/Accelerate/vImage_Buffer.swift:
```diff
@@ -0,0 +1,348 @@
+    public init?(size: CGSize,
+                 bitsPerPixel: UInt32,
+                 error: inout Int,
+                 flags options: vImage.Options = .noFlags) {
```
The only flags options supported by `vImageBuffer_Init` are `kvImagePrintDiagnosticsToConsole`, `kvImageNoAllocate`, and `kvImageNoFlags`. The first introduces a *third* mode of error handling to this `init` (you already have the error code as well as returning `nil`), and the second is half-baked for reasons described above. This init doesn't need to take a flags parameter, and I don't think you need the inout error param either. Simply return nil for error. That's the idiom for failable inits in Swift.

FlexMonkey replied:
I've made this throw - there are two potential errors, illegal dimensions or a `kvImageMemoryAllocationError`. Other initializers, such as `vImageBuffer_InitWithCGImage` can throw a much wider selection of errors, having all my buffer inits throwable gives a consistent API.

 I'm moving onto removing `flags` now...

FlexMonkey replied:
`flags` removed.

stephentyrone commented on stdlib/public/Darwin/Accelerate/vImage_Buffer.swift:
```diff
@@ -0,0 +1,348 @@
+    /// - Returns: An initialized vImage buffer.
+    public init?(size: CGSize,
+                 bitsPerPixel: UInt32,
+                 flags options: vImage.Options = .noFlags) {
```
I think that this is the only `init?` you should actually be exposing in the overlay, except that it probably shouldn't have `flags` (as noted above) and should probably take `width` and `height` as `Int` (unless there's a specific reason for `CGSize`).

FlexMonkey replied:
Flags have gone and `size` replaced by `width` and `height`.

stephentyrone commented on stdlib/public/Darwin/Accelerate/vImage_Buffer.swift:
```diff
@@ -0,0 +1,348 @@
+    ///
+    /// This function will instantiate and initialize a vImage buffer from a `CGImage` using a `CGImageFormat`
+    /// based on the provided image's properties.
+    public init?(cgImage: CGImage,
```
`_ cgImage: CGImage`; this is effectively a conversion, and shouldn't have a parameter label. You don't need `error`, because the init is failable. More generally, adopt this pattern throughout--either fail (trap or return nil) *or* return an error code, but not both. I'm not even sure that `flags` pulls its weight here--the only flag that's semantically valid in this function is `kvImageDoNotTile`.

FlexMonkey replied:
WRT to parameter name, I was taking my lead from `UIImage` that has initializers from [CG](https://developer.apple.com/documentation/uikit/uiimage/1624090-init) and [CI](https://developer.apple.com/documentation/uikit/uiimage/1624114-init) images, both with named parameters. 

I originally only had failable inits,  but I had pushback from engineering who wanted an option to get the error code back to facilitate figuring out why a particular initialization had failed.

`vImageBuffer_InitWithCGImage` also supports `kvImagePrintDiagnosticsToConsole`, and my `flags` parameter comes with a default.

stephentyrone replied:
If you want to produce a descriptive error, this should throw, rather than having multiple error options.

FlexMonkey replied:
Agreed. I've removed all instances of passing an inout error variable in favor of throwing a proper error.

stephentyrone commented on stdlib/public/Darwin/Accelerate/vImage_Buffer.swift:
```diff
@@ -0,0 +1,348 @@
+    public init?(cgImage: CGImage,
+                 format: vImage_CGImageFormat,
+                 error: inout Int,
+                 flags options: vImage.Options = .noFlags) {
```
As above, this signature should be:
`init?(_ cgImage: CGImage, format: vImage_CGImageFormat, flags: options: vImage.Options = .noFlags)`

FlexMonkey replied:
Fixed - made throwable.

stephentyrone commented on stdlib/public/Darwin/Accelerate/vImage_Buffer.swift:
```diff
@@ -0,0 +1,348 @@
+    /// - Returns: A Core Graphics image containing a representation of the vImage buffer.
+    public func createCGImage(format: vImage_CGImageFormat,
+                              error: inout Int,
+                              flags options: vImage.Options = .noFlags) -> CGImage? {
```
This would probably be better as an init on CGImage, but you should discuss that with the CG team.

FlexMonkey replied:
Email sent.

FlexMonkey replied:
Convenience initializers are not supported in extensions of CF types.

stephentyrone commented on stdlib/public/Darwin/Accelerate/vImage_Buffer.swift:
```diff
@@ -0,0 +1,348 @@
+            error =  vImageCopyBuffer($0,
+                                      &destinationBuffer,
+                                      rowBytes / Int(width),
+                                      options.flags)
```
This is unsafe in all sorts of ways that should be checked in the Swift wrapper; two off the top of my head:
- enforce that `destinationBuffer`s width and height are >= the source buffer's.
- rowBytes / width may produce zero for sub-byte pixel formats.

FlexMonkey replied:
I already have a check to prevent divide by zero:

        if Int(width) == 0 {
            throw vImage.Error(vImageError: kvImageInvalidParameter)
        }

I'm now throwing an error that mirrors any errors generated by vImage. So, I could add:

        if width < destinationBuffer.width || height < destinationBuffer.height {
            throw vImage.Error(vImageError: kvImageBufferSizeMismatch)
        }

...but the behavior is identical with or without it.

stephentyrone commented on stdlib/public/Darwin/Accelerate/vImage_CGImageFormat.swift:
```diff
@@ -0,0 +1,72 @@
+    }
+    
+    /// The number of color channels.
+    public var componentCount: UInt32 {
```
This should be `Int`.

FlexMonkey replied:
Absolutely. Fixed.

stephentyrone commented on stdlib/public/Darwin/Accelerate/vImage_CGImageFormat.swift:
```diff
@@ -0,0 +1,72 @@
+    }
+    
+    /// Initializes an image format.
+    public init(bitsPerComponent: UInt32,
```
These should be `Int` instead of `UInt32`.

FlexMonkey replied:
Ideally, yes. But the verbose original version of this init uses `UInt32` (see [here](https://developer.apple.com/documentation/accelerate/vimage_cgimageformat/1546793-init)). I kept the same type for my lightweight init for consistency.

stephentyrone replied:
I don't think "consistency of being a pain in the butt to use" is a very useful guiding principle for API design. =)

FlexMonkey replied:
Good point. Changed to `Int`.

stephentyrone commented on stdlib/public/Darwin/Accelerate/vImage_CGImageFormat.swift:
```diff
@@ -0,0 +1,72 @@
+//===----------------------------------------------------------------------===//
+
+@available(iOS 9999, OSX 9999, tvOS 9999, watchOS 9999, *)
+extension vImage_CGImageFormat {
```
If you're going to define `==`, you should conform to `Equatable`. Conforming to `Hashable` as well would be a good idea.

FlexMonkey replied:
The reason it's not `Equatable` is that protocol 'Equatable' requires '==' to be available on OS X 10.9.0 and newer, and we're specifying 9999. The reason it's not `Hashable` is "the components used for hashing must be the same as the components compared in your type’s == operator implementation" - `vImageCGImageFormat_IsEqual` is a black box, so I don't know which properties are used for equivalence. Potentially, I could try something like:

    public func hash(into hasher: inout Hasher) {
        hasher.combine(self.bitmapInfo.rawValue)
        hasher.combine(self.bitsPerComponent)
        hasher.combine(self.bitsPerPixel)
        hasher.combine(self.colorSpace.takeRetainedValue().hashValue)
        hasher.combine(self.componentCount)
        hasher.combine(self.renderingIntent.rawValue)
    }

...but `Hashable` has the same 10.9.0 requirement.

stephentyrone replied:
@airspeedswift This seems like an appropriate use for @_alwaysEmitIntoClient.

lorentey replied:
If it isn't well-defined what components `==` actually compares, then it makes sense not to implement `Hashable`. (Although: what's the value of providing equality if it's not clear what it does?) The risk here is that the lack of an official `Hashable` conformance is inviting API users to try conforming this type to `Hashable` themselves.

An alternative would be to implement both `==` and `hash(into:)` from scratch here, not calling into `vImageCGImageFormat_IsEqual` at all.

FlexMonkey replied:
@stephentyrone - I have a query outstanding to understand the mechanics of `vImageCGImageFormat_IsEqual`. Let me know if you think it's worth removing this function in the short term and reintroduce it when we have properly understood equivalence and can implement `Hashable`.

stephentyrone replied:
Yes, let's do that.

FlexMonkey replied:
Super. That's removed and tests pass locally. Cheers!

FlexMonkey committed:
`vImage_CGImageFormat.componentCount` should be `Int`.

FlexMonkey committed:
`vImage_CGImageFormat.componentCount` should be `Int`

FlexMonkey committed:
Move `self.init()` to after the size check for kvImageNoAllocate init.

FlexMonkey committed:
Buffer initializers to width and height rather than size.

FlexMonkey committed:
change vImage_CGImageFormat lightweight initializer to accept Int for `bitsPerComponent` and `bitsPerPixel`.

FlexMonkey committed:
Flesh out docs for vImage_Buffer.size

FlexMonkey committed:
Remove faux initializer in favor of new static function: `preferredAlignmentAndRowBytes`.

Commit:
Merge branch 'master' into accelerate-vImage

Commit:
Change functions to use proper error handling rather than inout error codes.

Commit:
Removed `flags` from basic init.

The only real flag to pass here is print diagnostics to console, and I now throw proper errors.

Commit:
Tests to check error throwing for buffer copy.

Commit:
remove unnecessary import, add missing docs.

Commit:
Add comments to error enums.

Commit:
Merge branch 'master' into accelerate-vImage

stephentyrone commented:
@swift-ci please test

stephentyrone commented:
@swift-ci please test source compatibility

FlexMonkey committed:
Merge branch 'master' into accelerate-vImage

FlexMonkey committed:
Fix bug creating string from format code.

stephentyrone commented:
@swift-ci please test

stephentyrone commented on stdlib/public/Darwin/Accelerate/vImage_CVImageFormat.swift:
```diff
@@ -134,12 +134,22 @@ extension vImageCVImageFormat {
+                String($0)
+            }.forEach {
+                returnString.append($0)
+        }
```
What's the benefit of exposing this property as a `String`? How is a developer going to use this in practice, and is String really preferred to UInt32 for that purpose?

FlexMonkey replied:
Good point, I've changed back to simply returning the `UInt32`.

lorentey commented on stdlib/public/Darwin/Accelerate/vImage_CVImageFormat.swift:
```diff
@@ -0,0 +1,659 @@
+    }
+    
+    /// The format code of a Core Video image format.
+    public var formatCode: String {
```
I think these format codes would be better represented by a dedicated wrapper type around an `UInt32` value. Conversion between `String` and the type code representation is laborious.

When we do convert these to strings, we should make sure we're creating a small (inline) Swift string, in a consistent and efficient way. `UTCreateStringForOSType` incurs allocation, transcoding and bridging overhead, and its results are inconsistent with the values generated on non-macOS platforms (it uses Mac OS Roman, not Latin-1).


```
@_fixed_layout
public struct vImageFormatCode: RawRepresentable {
  @usableFromInline
  internal let rawValue: UInt32

  @inlinable
  public init(rawValue: UInt32)

  public var stringRepresentation: String {
    // FIXME: This uses Latin-1, not MacRoman.
    var result = ""
    result.append(String(UnicodeScalar(formatCode >> 24 & 0xFF)!))
    result.append(String(UnicodeScalar(formatCode >> 16 & 0xFF)!))
    result.append(String(UnicodeScalar(formatCode >> 8 & 0xFF)!))
    result.append(String(UnicodeScalar(formatCode >> 0 & 0xFF)!))
    return result
  }
}

extension vImageCVImageFormat {
  public formatCode: vImageFormatCode {
    let formatCode = vImageCVImageFormat_GetFormatCode(cvConstImageFormat)
    return vImageFormatCode(formatCode)
  }
}
```

FlexMonkey replied:
See above, this property is now a `UInt32`.

Commit:
Make `vImageCVImageFormat.formatCode` a `UInt32`.

Commit:
Remove equivalence operator from `CGImageFormat`.

stephentyrone commented:
@swift-ci please smoke test

stephentyrone commented:
@swift-ci please test source compatibility
